api/cluster.go

```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
 // name:         name to be used in logging & tracing
 // path:         path to request on
 func (s *Server) peerQuerySpeculative(ctx context.Context, data cluster.Traceable, name, path string) (map[string]PeerResponse, error) {
-	peerGroups, err := cluster.MembersForSpeculativeQuery()
-	if err != nil {
-		log.Error(3, "HTTP peerQuery unable to get peers, %s", err)
-		return nil, err
-	}
-	log.Debug("HTTP %s across %d instances", name, len(peerGroups)-1)
+	result := make(map[string]PeerResponse)
 
-	reqCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	responseChan, errorChan := s.peerQuerySpeculativeChan(ctx, data, name, path)
 
-	originalPeers := make(map[string]struct{}, len(peerGroups))
-	receivedResponses := make(map[int32]struct{}, len(peerGroups))
+	for resp := range responseChan {
+		result[resp.peer.GetName()] = resp
+	}
 
-	responses := make(chan struct {
-		shardGroup int32
-		data       PeerResponse
-		err        error
-	}, 1)
+	err := <-errorChan
+	return result, err
+}
 
-	askPeer := func(shardGroup int32, peer cluster.Node) {
-		log.Debug("HTTP Render querying %s%s", peer.GetName(), path)
-		buf, err := peer.Post(reqCtx, name, path, data)
+// peerQuerySpeculativeChan takes a request and the path to request it on, then fans it out
+// across the cluster. If any peer fails requests to other peers are aborted. If enough
+// peers have been heard from (based on speculation-threshold configuration), and we
+// are missing the others, try to speculatively query other members of the shard group.
+// ctx:          request context
+// data:         request to be submitted
+// name:         name to be used in logging & tracing
+// path:         path to request on
+// resultChan:   channel to put responses on as they come in
+func (s *Server) peerQuerySpeculativeChan(ctx context.Context, data cluster.Traceable, name, path string) (<-chan PeerResponse, <-chan error) {
+	resultChan := make(chan PeerResponse)
+	errorChan := make(chan error, 1)
 
-		select {
-		case <-ctx.Done():
-			return
-		default:
-			// Not canceled, continue
-		}
+	go func() {
```

Comment:
I think this  func is like a lambda?  Its kind of long.  Would there be value in refactoring into smaller methods called from the lambada?

(although it looks like longer functions are the norm in the code base)

Reply:
It is like a lambda, and behaves as a closure (captures scope vars that are referenced). It may be worth refactoring? not sure.